### PR TITLE
Balance bionic power use for realistic values

### PR DESCRIPTION
**Pull request by RDru**
#### Summary
```SUMMARY: Balance "Balance bionic power use for realistic values"```

#### Purpose of change
We have realistic units and lower bionic power now.

#### Describe the solution

1W = 1J/1s

2W for "databanks": real SSD is 2W or less
2W for internal processors: ARM processors and smartphones use around 2W or less (https://www.anandtech.com/show/13686/snapdragon-855-power-consumption-but)

- Adrenaline Pump: 5J
"stimulator system has been surgically implanted alongside your adrenal glands" should not need much
- Alarm System: 1W
motion detector in real: 0.5 - 1W
- Monomolecular Blade: 50J
just some engine for extending and hiding the blade
- Shotgun Arm: 50J
just some engine for extending and hiding the gun
- Blood Analysis: 100J
- Blood Filter: 600J
I'm guessing 10W filter (micro pump + microchip). It takes less than a minute for blood to circulate the body, so 10x60 single use
- Chain Lightning: 100000J
"The current world's largest Tesla coil is a 130,000-watt": https://en.wikipedia.org/wiki/Tesla_coil
- Bionic Claws: 25J
small engine for extending and hiding the claws
- Internal Climate Control: 100W
typical fridge uses 100W when active, internally 100W should also be enough for heating, and we don't always need to cool or heat
- Cloaking System: 30000W
"This high-power system uses a set of cameras and LEDs to make you blend into your background" this should not work with too much armor/clothing but should not drain much power otherwise. For now I'm changing this to 30000W (same proportion with night generator as before) for balance reasons (it's superior).
- Close Quarters Battle: 20W
"Bionic processors and databanks (...) are surgically integrated into your nervous system."
2W for databank, 2W for processor. I'm going for 5 total sets, each limb + central.
- Directional EMP: 50000J
http://www.thespacereview.com/article/1549/2: "EMP effects upon IC-based devices" "Tests with EMP simulators have shown that a very short pulse of about 10-7 Joule is sufficient to damage a microwave semiconductor diode, and roughly .05 J will damage an audio transistor, whereas 1 J would be required for vacuum tube damage" "A few watts to a few hundred watts of power are sufficient to destroy most ICs, when delivered in a few nanoseconds (e.g. 10-7 J /10-8 sec = 10 W)." I'm going with 50 000W since we don't fry ICs directly but through armor/shielding of turrets and robots.
- EMP Projector: 50000J
- Aero-Evaporator: 720000J
real driers use up to 1000W, they make 60L per day, that's 2,5L/h. Using bionic in game can give 0,5L. That's 12min of use. 1000x60x12. That's 720kJ, 720 old units, but only 3/4 of heavy battery. Could be made into constant use with small water container we can drink or siphon from.
- Dielectric Capacitance System: 5W
If going for realism this should give power not use it. Setting 5W for standby use.
- Flashbang Generator: 10000J
few kW leds + few kW speakers should do the trick
- Cranial Flashlight: 3W
typycil led flashlight
- LED Tattoo: 1W
- Integrated Dosimeter: 100J
- Terranian Sonar: 200W
small ground-penetrating radar like this should only use 50W+
- Thermal Dissipation: 1000W
"Powerful heatsinks and supermaterials" - if heat is dissipated through that then 1kW should be more than enough
- Hydraulic Muscles: 10000W
https://www.quora.com/How-do-I-calculate-the-electric-power-required-to-lift-100-kg-of-material: 1kW is enough to lift 100kg per 1m per 1s. 20 str is 200kg lift in game. Efficiency 20%.
- Infrared Vision: 5W
3-5W real
- Finger-Mounted Laser: 30000J
https://en.wikipedia.org/wiki/AN/SEQ-3_Laser_Weapon_System: going for "(30 kW) to fry sensors, burn out motors, and detonate explosive materials"
- Leukocyte Breeder System: 5W
- Mini-Flamethrower: no change
how to calculate power to fire? this should probably use fuel
- Fingerpick: 50J
micro motors working for few seconds
- Electromagnetic Unit: 10000W
https://www.quora.com/How-much-electricity-does-a-powerful-electromagnet-consume: "8.5kW to 15 kW" for scrap magnet
- Enhanced Memory Banks: 2W
based on real SSD drives( 2W on load, 0.5 on idle), plus some extra microcontroller
- Weather Reader: 10J
1W use, lets say we need 10s for it to work
- Artificial Night Generator: 9000W
this works by destructive interference. So we need same light wave as source but shifted. Same wave = same power needed. We need this to be powerful enough to block the sun. https://ag.tennessee.edu/solar/Pages/What%20Is%20Solar%20Energy/Sun%27s%20Energy.aspx Sun's power is 1000W per square meter. It blocks in 15 tile radius, that's 29x29 square, 841 m2, way too much. I'm changing this to 3x3 square. Should be changed to dynamic range or dynamic power use.
- Implanted Night Vision: 10W
- Offensive Defense System: 10000W
Tesla / 10. It's same as before.
- Probability Travel: 3W 
it's standby power drain
- Radiation Scrubber System: 4500J
no idea, so let's keep the old ratio compared to blood filters
- Remote Controller: 10W
10W for strong signal, probably too much
- Scent Vision: 2W
let's assume it's a micro computer connected to your brain or eyes
- Uncanny Dodge: 6W
"Your nervous system has been augmented with bionic processors", again assuming 5 processors - each limb + central/head
- Water Extraction Unit: 20000W
typical garden water pump uses around 100W, how fast can you pump water from bodies, 20kW for balance. Should be changed to take time.

#### Describe alternatives you've considered
Just playing the game.

#### Additional context
Bionic UI doesn't show values other than kJ. We need to show lower values.

### CODE REVIEW

File: src/lightmap.cpp [hunk 1]
```diff
@@ -447,7 +447,7 @@ void map::generate_lightmap( const int zlev )
 
     if( g->u.has_active_bionic( bionic_id( "bio_night" ) ) ) {
         for( const tripoint &p : points_in_rectangle( cache_start, cache_end ) ) {
-            if( rl_dist( p, g->u.pos() ) < 15 ) {
+            if( rl_dist( p, g->u.pos() ) < 2 ) {
```
Comment (Fris0uman) [inline src/lightmap.cpp]: Why change this?

Comment (RDru) [reply src/lightmap.cpp]: Please read about Artificial Night Generator in PR's description

Comment (Fris0uman) [reply src/lightmap.cpp]: Oh I missed your explanation sorry.

Looking at your math couldn't you make the power draw 852 kJ?

Comment (RDru) [reply src/lightmap.cpp]: We have only a little over 2000 kJ available. What would be the point of that bionic if it's unusable?

Comment (RDru) [reply src/lightmap.cpp]: Also power draw that high is unrealistic for any normal device.

**Comment by snipercup** [comment]
>     * Implanted Night Vision: 10W

I don't see any change in the implated night vision. Can you maybe change the description of your PR to make it clear what changes and what does not?

**Comment by RDru** [comment]
> 
> 
> > ```
> > * Implanted Night Vision: 10W
> > ```
> 
> I don't see any change in the implated night vision. Can you maybe change the description of your PR to make it clear what changes and what does not?

Isn't it clear? Everything in description is changed unless stated otherwise.

**Comment by Night-Pryanik** [comment]
> 20 str is 50kg lift in game.

This seems wrong. We assume a character can lift 10 kg per 1 point of strength.

**Comment by mrkybe** [comment]
> Uncanny Dodge: 100W
> "Your nervous system has been augmented with bionic processors", 100W in brain sound like a lot?

It is a lot.  [The brain uses, apparently, 20W](https://hypertextbook.com/facts/2001/JacquelineLing.shtml)  I'm pretty sure adding that much extra heat to the brain would be very fatal.  Bionic processors are probably notable for being lower power.  I would say like, 5W passive, possibly spiking when it is actually used?

**Comment by Zireael07** [comment]
Add me to the camp that thinks 100W would be fatal. I remember a sci-fi or a cyberpunk story saying 400 mA is enough to kill.

**Comment by reed501** [comment]
What does "10W instant" mean? Not 10 specifically but an instant wattage. A Watt is energy per second and you store energy, does this mean it drains at 10W for an instantaneous amount of time? If so that would mean 10J, unless I'm misunderstanding how this is described.

**Comment by RDru** [comment]
Yes, it means 10J :)
Will change that and work on power drain of motors and cpus later.

**Comment by RDru** [comment]
I have done a revision of all bionics.
Went with 2W internal processors. 50J for blade extending, 25J for claws.
Check commits and PR's description of why I did what.

**Comment by kevingranade** [comment]
> This seems wrong. We assume a character can lift 10 kg per 1 point of strength.

This is the status quo, but it's also crazy wrong, it's at least twice what it should be.

**Comment by Night-Pryanik** [comment]
> This is the status quo, but it's also crazy wrong, it's at least twice what it should be.

Why? An average survivor with default 8 strength should be able to lift only 40 kilograms?

**Comment by kevingranade** [comment]
I was thinking about carrying capacity, which is 4kg per point of strength with no penalty, which is the number I consider to be about 2x too high.
The maximum a player can carry in inventory is actually 16kg per point of strength, which is similarly ludicrous.
I'm assuming the 10kg per point is for something else, like lifting a single item,  haven't looked into that.

**Comment by Night-Pryanik** [comment]
> I'm assuming the 10kg per point is for something else, like lifting a single item, haven't looked into that.

Yeah, exactly this.

**Comment by RDru** [comment]
I'm confused and don't know if I should change anything :)

File: data/json/bionics.json [hunk 2]
```diff
@@ -1041,8 +1041,8 @@
     "description": "Your nervous system has been augmented with bionic processors, allowing you to dodge attacks beyond normal human capability, including bullets.",
     "occupied_bodyparts": [ [ "TORSO", 12 ], [ "ARM_L", 3 ], [ "ARM_R", 3 ], [ "LEG_L", 5 ], [ "LEG_R", 5 ], [ "FOOT_L", 1 ], [ "FOOT_R", 1 ] ],
     "flags": [ "BIONIC_TOGGLED" ],
-    "act_cost": 1,
-    "react_cost": 1,
+    "act_cost": "6 J",
+    "react_cost": "6 J",
     "time": 10
```
Comment (Aloxaf) [inline data/json/bionics.json]: You said Uncanny Dodge is 6W, so it may should be set to 1 here?
```suggestion
     "time": 1
```

**Comment by justamblingalong** [comment]
It seems like a problem to try to implement/balance a system where you have such a huge range of power values. I can't imagine how this wouldn't translate into either a) energy-intense modules like Chain Lightning being effectively impossible to use or b) energy-cheap modules like IF vision being effectively free.

If players are going to be using 100,000+ J of energy for some bionics, it seems needlessly kludgy to then force them to track bionics that are consuming 1-2J at a time (i.e. .001% or less their storage).

Either some level of normalization is needed, or alternatively, very cheap CBMs should be made freely toggleable.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/is-the-bionics-nerf-intentional/21601/2

**Comment by mrkybe** [comment]
> It seems like a problem to try to implement/balance a system where you have such a huge range of power values. I can't imagine how this wouldn't translate into either a) energy-intense modules like Chain Lightning being effectively impossible to use or b) energy-cheap modules like IF vision being effectively free.

Maybe it would make sense to split bionic power into two types?

Battery Banks and Capacitive System CBM.

Battery power would function, more or less, how it currently does.  It would be used for powering active CBMs that consume some amount of power every turn they're active.

Capacitive power would be what is drained to turn systems on or off, or fire CBMs that require large spikes of energy.

Capacitors would be charged from Battery Banks, but unlike them, would discharge at a rate proportional to the amount of charge stored in them already, (possibly differences between civilian grade ones and military grade ones?) and like in real life, have a lower energy density compared to batteries.

They would only be charged on demand by the CBM you activate, or if activated by the player, the Capacitive CBM would begin to charge to full power.  The consumed power is basically lost if not used somehow.  Kind of like a 'prepare for combat' type thing.

So for example, to fire Chain Lightning, you'd need at least a Military Grade version of the Capacitor System CBM.  Activating Chain Lighting would discharge all the energy in the capacitor, and do damage proportional to the amount of energy discharged.

IIRC, there's already a Capacitive System CBM that's used to become invulnerable to lightning?  Ideally this would supplant it, allowing lighting to charge your capacitors, but only until they're full, at which point it starts to do damage again.

**Comment by RDru** [comment]
I was actually thinking about splitting bionics into components. Not only batteries and capacitors, but also things like processors, memory banks, fuel storage, software. 
Let's say you find some bionic with processors and memory banks in a body. You take them but some are damaged, software is damaged. You need new software or you can maybe increase efficiency of what you already have in you.
Anyway, this is out of scope of this PR's.

**Comment by DaviBones** [comment]
>Hydraulic Muscles: 2500W
https://www.quora.com/How-do-I-calculate-the-electric-power-required-to-lift-100-kg-of-material: 1kW is enough to lift 100kg per 1m per 1s. 20 str is 200kg lift in game. Setting for 2kW + 500W for walking, friction.

Complex hydraulic systems are not that energy efficient. They start out pretty efficient, but each time you have to re-direct the hydraulic flow you multiply that efficiency by 80% to 95% depending on your price point, and that's for joints that don't move, never mind hinge joints (elbow) or ball joints (shoulder). 

According to a study mentioned in [this blog post](https://www.hydraulicspneumatics.com/blog/how-efficient-are-your-hydraulic-machines), the average energy efficiency of fluid power systems in 2014 was 21%. I think that would be a decent estimate to start from. That would change the numbers to 2kW used + 8kW wasted for 10kW total.

**Comment by RDru** [comment]
Setting for 10k then

**Comment by vlad1492** [comment]
Applause!  Glad to see bionics getting a workover and more physics respect. 

Quibble: 

Night generator doesn't need to counter entire spectrum, only the visible. Perhaps lower the power requirement or extend range some.

**Comment by Pmm5000** [comment]
Few suggestions right quick. 

What is the power draw of repair nanobots? I think it is still 5 kJ? 
Looking at this, https://worldbuilding.stackexchange.com/questions/35975/how-do-you-power-all-the-nanobots , nominal power seems to be 2 W, but would also cause severe issues should they ever not receive power after about 4 hours. Is there any way to kinda do the same thing like the breeder system?

Then I also imagine the Olfactory Mask would have similar power draw to Uncanny Dodge 6 W since it affects the whole body.

Remote Controls also use a pretty negligible amount of power. I could see it however using 1 W every 24 turns. 

Then I feel like the integrated tool set is a weird one. 
Since we're looking at more realistic values: 
as a soldering iron, it uses up somewhere between 20 W-60 W

as a welding tool, https://www.quora.com/What-is-the-wattage-of-single-phase-welding-machine , it apparently uses 20 KW at it's maximum usage, although looking at https://en.wikipedia.org/wiki/Welding_power_supply
W=VxA and according the that the Voltage range is 17 V-45 V and the Amperage range is 55 A - 590 A, therefore the hypothetical range is 935 W- 26550 W, or roughly 1 kJ - 26 kJ. However this doesn't factor in total time of each activity. I feel like I'm getting lost now. Darn. 

as for it's manual toolsets, it would probably use a small engine like the bionic claws, so 25 J.

**Pull request merged**